stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -243,8 +231,10 @@ public void onHeaders(Metadata.Headers headers) {
     public void onPayload(RespT payload) {
       observer.onValue(payload);
 
-      // Request delivery of the next inbound message.
-      call.request(1);
+      if (streamingResponse) {
```

Comment:
If there are more messages, we need to fail. If you put in this optimization, then I think you also have to make sure to `request(2)` at the beginning and to fail appropriately if more than one payload is received.